docs/specifications/4-standard-compliance-suite.md

```diff
@@ -0,0 +1,500 @@
+    switch (keys.size) {
+      case 1:
+        return keys.has('offeredTo');
+      case 2:
+      return keys.has('offeredTo') && keys.has('makesYouForgetTheMatrix');
+      default:
+        return false;
+    }
+  }
+  public static isRed(pill: BluePill | RedPill): pill is RedPill {
+    const keys = new Set(Object.keys(pill));
+    switch (keys.size) {
+      case 1:
+        return keys.has('offeredTo');
+      case 2:
+        return keys.has('offeredTo') && keys.has('makesYouExitTheMatrix');
+      default:
+        return false;
+    }
+  }
+  private constructor(){}
+}
+export class Neo {
+  public readonly tookBlue: boolean;
+  public readonly tookRed: boolean;
+
+  public constructor(public readonly pill: BluePill | RedPill) {
+    this.tookBlue = pill.offeredTo == 'Neo' && Morpheus.isBlue(pill);
+    this.tookRed = pill.offeredTo == 'Neo' && Morpheus.isRed(pill);
+  }
+}
+
+// WHEN
+const bluePillA = new Neo({ offeredTo: 'not Neo' });
+const bluePillB = new Neo({ offeredTo: 'Neo', makesYouForgetTheMatrix: true });
+const redPillA = new Neo({ offeredTo: 'not Neo' });
+const redPillB = new Neo({ offeredTo: 'Neo', makesYouExitTheMatrix: true });
+
+// THEN
+expect(bluePillA.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(bluePillB.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeTruthy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillA.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillB.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeTruthy();
+```
+
+##### Kernel Trace
+```
+```
+</details>
+
+#### Partially initialized object consumption
+When a constructor passes `this` out from **JavaScript** to the *host* app, the
+reference must be correctly identified and passed across.
+
+> :construction: The .NET Runtime does not currently honor object identity,
+> meaning that despite the same object reference is returned twice, two distinct
+> proxies exist for it in the *host* .NET app.
+
+<details><summary>Show test</summary>
+
+##### Reference Implementation
+```ts
+// GIVEN
+export abstract class PartiallyInitializedThisConsumer {
+  public abstract consumePartiallyInitializedThis(obj: ConstructorPassesThisOut): void;
+}
+export class ConstructorPassesThisOut {
+  public constructor(consumer: PartiallyInitializedThisConsumer) {
+    consumer.consumePartiallyInitializedThis(this);
+  }
+}
+
+// WHEN
+class MyConsumer extends PartiallyInitializedThisConsumer {
+  public obj?: ConstructorPassesThisOut = null;
+
+  public consumePartiallyInitializedThis(obj: ConstructorPassesThisOut) {
+    this.obj = obj;
+  }
+}
+const consumer = new MyConsumer();
+const object = new ConstructorPassesThisOut(consumer);
+
+// THEN
+expect(consumer.obj).toBe(object);
```

Comment:
So rather than saying "C# fails this test", we don't have this test, I suppose?

Or the test should be about observable side effects, not identity. Changes made to the object that is passed out are visible via the other reference as well, and so we know they "must be" the same object, or at least behave as the same object, regardless of the strict object identity.

Reply:
This is a case where they *could* (and in fact, *should*) be the same object. I've intentionally left this test here (in this state) because I think this particular predicate (same reference with same static type must be the same object instance on the host) is valuable, and is definitely feasible.

I want to go with it this way for a while and possibly revisit if we find out the benefits is not worth the cost (I *guess* the cost isn't very high now... but I could guess wrong!)